csharp-advanced-topics/VolatileInterlockedLockInCsharp/VolatileInterlockedLockInCsharp/Account.cs

```diff
@@ -0,0 +1,70 @@
+﻿namespace VolatileInterlockedLockInCsharp
+{
+    public class Account
+    {
+        private int _balance;
+        private volatile int _balanceVolatile;
+        private int _balanceLock;
+        private int _balanceInterlocked;
+        private object _lock = new object();
+
+        public int Balance
+        {
+            get => _balance;
+            set => _balance = value;
+        }
+
+        public int BalanceVolatile
+        {
+            get => _balanceVolatile;
+            set => _balanceVolatile = value;
+        }
+
+        public int BalanceLock
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    return _balanceLock;
+                }
+            }
+            set
+            {
+                lock (_lock)
+                {
+                    _balanceLock = value;
+                }
+            }
+        }
+
+        public int BalanceInterlocked
+        {
+            get => _balanceInterlocked;
+            set => _balanceInterlocked = value;
+        }
+
+        public void Withdraw(int amount)
+        {
+            _balance -= amount;
+        }
+
+        public void WithdrawVolatile(int amount)
+        {
+            _balanceVolatile -= amount;
+        }
+
+        public void WithdrawLock(int amount)
+        {
+            lock (_lock)
+            {
+                _balanceLock -= amount;
+            }
+        }
+
+        public void WithdrawInterlocked(int amount)
+        {
+            Interlocked.Exchange(ref _balanceInterlocked, _balanceInterlocked - amount);
+        }
```

Comment:
While this may be hard to reproduce, especially the way the code is written, this code here has a race condition. The interlocked exchange will guarantee atomic write access to the `ref` value, but we are still reading the `_balanceInterlocked` value without any synchronization. If you look at the JIT output of that call, you will see an `lea` call on the `_balanceInterlocked` value. Followed by `add` and finally the `xchg` call, which is a full fenced atomic operation.
We can solve this by using `Interlocked.Add(ref _balanceInterlocked, -amount)`